jakarta-eclipse/src/org/jakarta/jdt/PersistenceEntityDiagnosticsCollector.java

```diff
@@ -0,0 +1,181 @@
+					
+					ISourceRange nameRange = JDTUtils.getNameRange(type);
+					Range range = JDTUtils.toRange(unit, nameRange.getOffset(), nameRange.getLength());
+					
+					/*============ Entity Annotation Diagnostics ===========*/
+					IAnnotation EntityAnnotation = null;
+					for (IAnnotation annotation: allAnnotations) {
+						if (annotation.getElementName().equals(PersistenceConstants.ENTITY)) {
+							EntityAnnotation = annotation;
+						}
+					}
+					
+					if (EntityAnnotation != null) {
+						
+						ISourceRange annotationNameRange = JDTUtils.getNameRange(EntityAnnotation);
+						Range annotationrange = JDTUtils.toRange(unit, annotationNameRange.getOffset(), annotationNameRange.getLength());
+										
+						// Define boolean requirements for the diagnostics
+						boolean hasPublicOrProtectedNoArgConstructor = false;
+						boolean isEntityClassFinal = false;
+						boolean isMethodsOrPersistentVariablesFinal = false;
+						
+						// Get the Methods of the annotated Class
+						for (IMethod method: type.getMethods()) {
+							if (method.isConstructor()) {
+								// We have found a method that is a constructor
+								if (method.getNumberOfParameters() > 0) continue;
+								
+								// Don't need to perform subtractions to check flags because eclipse notifies on illegal constructor modifiers
+								if (method.getFlags() != Flags.AccPublic && method.getFlags() != Flags.AccProtected) continue;
+								
+								hasPublicOrProtectedNoArgConstructor = true;
+							}
+							
+							// All Methods of this class should not be final
+							if (this.isFinal(method.getFlags())) {
+								diagnostics.add(createDiagnostic(method, unit, "Class with @Entity annotation cannot contain any final methods or persistent instance variables."));
+								isMethodsOrPersistentVariablesFinal = true;
+							}
+						}
+						
+						// Go through the instance variables and make sure no instance vars are final
+						for (IField field: type.getFields()) {
+							
+							// If a field is static, we do not care about it, we care about all other field
+							if(this.isStatic(field.getFlags()))
+							{
+								continue;
+							}
+							
+							// If we find a non-static variable that is final, this is a problem
+							if (this.isFinal(field.getFlags())) {
+								diagnostics.add(createDiagnostic(field, unit, "Class with @Entity annotation cannot contain any final methods or persistent instance variables."));
+								isMethodsOrPersistentVariablesFinal = true;
+							}
+							
+						}
+						
+						// Ensure that the Entity class is not given a final modifier
+						if (this.isFinal(type.getFlags())) isEntityClassFinal = true;
+						
+						
+						// Create Diagnostics if needed
+						if (!hasPublicOrProtectedNoArgConstructor) {
+							diagnostics.add(createDiagnostic(EntityAnnotation, unit, "Class with @Entity annotation must contain a public or protected constructor with no arguments."));
```

Comment:
This diagnostic should be on the class name, and should read:
"A class using the`@Entity` annotation must contain a public or protected constructor with no arguments."